### PR TITLE
Uniffy tabs component

### LINKED ISSUE #4035: RemoteTabsProvider.setLocalTabs fails with `PoisonError` after unrelated sync failures
(sorry @Amejia481, I'm going to edit your initial bug comment here to add ongoing context; -@rfk)

## Current Status

As described in [this comment](https://github.com/mozilla/application-services/issues/4035#issuecomment-851102193), we think the issue here is that `SyncManager::do_sync` holds a `MutexGuard` for the tabs-store and logins-store for the duration of a sync. Any uncaught panic in the course of a sync will poison these mutexes, rendering the stores unusable even if the original error had nothing to do with tabs or logins.

In addition to fixing the underlying issues causing a panic during a sync (e.g. ref #4153) we should refactor things inside `SyncManager` so that it only holds the mutex for the tab store *while it is syncing tabs*, not for the whole duration of the sync.

------

The issue was addressed on AS `61.0.7` but it's happening again on AS `74.0.1`.

Sentry Fenix nightly https://sentry.prod.mozaws.net/operations/firefox-nightly/issues/8825532/
Fenix issue https://github.com/mozilla-mobile/fenix/issues/12162

image

**Stacktrace:**

```
mozilla.appservices.remotetabs.RemoteTabProviderException: called `Result::unwrap()` on an `Err` value: "PoisonError { inner: .. }"
    at mozilla.appservices.remotetabs.rust.RustError.intoException(RustError.kt:4)
    at mozilla.appservices.remotetabs.RemoteTabsProvider.setLocalTabs(RemoteTabsProvider.kt:11)
    at mozilla.components.browser.storage.sync.RemoteTabsStorage$store$2.invokeSuspend(RemoteTabsStorage.kt:15)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:3)
    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:15)
    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:1)
    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:11)
```

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/SYNC-2624)
┆Epic: Uniffi-cation of logins, tabs and history metadata
┆Fix Versions: Release 92
┆Sprint End Date: 2021-07-23

### PR DESCRIPTION
**Pull request by grigoryk**
(@rfk edit - Fixes #4035)

At the moment this is just a quick, unfinished stab at uniffying the tabs component.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/tabs/src/storage.rs [hunk 1]
```diff
@@ -8,7 +8,40 @@ const URI_LENGTH_MAX: usize = 65536;
...
+    //        to wrap the RefCell as they aren't `Sync`
+    // RefCell: So we can replace what it holds. Normally you'd use `get_ref()`
+    //          on the mutex and avoid the RefCell entirely, but that requires
+    //          the mutex to be declared as `mut` which is apparently
+    //          impossible in a `lazy_static`
+    // [Arc/Weak]<StoreImpl>: What the sync manager actually needs.
+    pub static ref STORE_FOR_MANAGER: Mutex<RefCell<Weak<StoreImpl>>> = Mutex::new(RefCell::new(Weak::new()));
+}
+
+// TODO how to avoid this local type?
```
Comment (rfk) [inline components/tabs/src/storage.rs:24]: Hrm, yeah, this is kind of unfortunate; we need to work towards being able to share types across uniffied components we we don't have any concrete ideas on how to make that work just yet :-(

**Review by rfk — COMMENTED** [review]
Neat, thanks for sharing! A couple of comments below from a first pass.

File: components/tabs/src/lib.rs [hunk 2]
```diff
@@ -15,11 +15,20 @@ pub mod msg_types {
...
 // `ClientRemoteTabs` structs without importing `sync15`.
-pub use sync15::clients::DeviceType;
+// pub use sync15::clients::DeviceType;
+
+use ffi_support::ConcurrentHandleMap;
+use std::sync::{Arc, Mutex};
+
+lazy_static::lazy_static! {
+    pub static ref STORES: ConcurrentHandleMap<Arc<Mutex<TabsStore>>> = ConcurrentHandleMap::new();
+}
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/tabs/src/lib.rs:32]: The best example of this is probably [here](https://github.com/mozilla/application-services/blob/fac0e27c78f37286f695369e41f29003f127a727/components/autofill/src/db/store.rs#L24)

File: components/tabs/src/tabs.udl [hunk 3]
```diff
@@ -0,0 +1,29 @@
+namespace tabs {
+
+};
+
+// TODO should this be [Threadsafe]? this runs into problems, e.g. mem_cached_state: Cell<MemoryCachedState> on TabsStorage
```
[the remark this answers is not in the document]
Comment (mhammond) [reply components/tabs/src/tabs.udl:5]: I'm inclined to suggest that we drop the ability to have a component perform "stand alone" syncs anyway - except maybe for testing (ie, an example or utility). So maybe consider deleting the `/// A convenience wrapper around sync_multiple.` helper entirely (and the `Cell` can die with it)

You will probably want a mutex around `storage` - and that's likely to mean you can also kill the `RefCell`s in the TabStorage struct.

**Review by mhammond — COMMENTED** [review]
lovely 👍

**Comment by mhammond** [comment]
> You will probably want a mutex around storage - and that's likely to mean you can also kill the RefCells in the TabStorage struct.

To be clear, because it took me an embarrassingly long time to discover this - once you have a mutex around the object, you can take a mutable reference to it, even from a function that doesn't have a `mut self` - so you can then just update, say, `local_tabs` just by assigning to it and avoiding the `RefCell` dance

**Comment by mhammond** [comment]
(possibly a `RwLock` is more idiomatic, but it doesn't really matter IMO)

**Comment by mhammond** [comment]
Via #4035 Ryan and I have concluded that because the sync manager holds the mutex around the tab store for the entire sync, any panic in any sync engine will cause the tabs mutex to be poisoned. Thus, I believe this PR will actually have a positive impact on reliability of the tabs component, so we should treat this as more important than a simple "nice to have".

**Comment by mhammond** [comment]
with grisha's blessing sammy took this over in #4192

**Pull request closed without merging**